addons/auth_signup/models/res_users.py

```diff
@@ -224,6 +239,71 @@ def send_unregistered_user_reminder(self, after_days=5):
             template = self.env.ref('auth_signup.mail_template_data_unregistered_users').with_context(dbname=self._cr.dbname, invited_users=invited_users[user])
             template.send_mail(user, email_layout_xmlid='mail.mail_notification_light', force_send=False)
 
+    def _should_alert_new_device(self):
+        if self._mfa_type() is None:
+            return True
+        if self._mfa_type() != 'totp_mail':
+            cookies = request.httprequest.cookies
+            key = cookies.get('td_id')
+            if key:
+                is_verified_device = request.env['auth_totp.device']._check_credentials_for_uid(
+                    scope="browser", key=key, uid=self.id)
+                if not is_verified_device:
+                    return True
+        return False
+
+    def _alert_new_device(self):
+        for user in self.env.user:
+            if user.email:
+                template = self.env.ref('auth_signup.alert_new_device_email')
+                assert template._name == 'mail.template'
+                email_values = {
+                    'email_cc': False,
+                    'auto_delete': False,
+                    'message_type': 'user_notification',
+                    'recipient_ids': [],
+                    'partner_ids': [],
+                    'scheduled_date': False,
+                }
+                context = self._prepare_new_device_notice_values()
+                email_values['email_to'] = user.email
+                template.with_context(context).send_mail(user.id, raise_exception=True, email_values=email_values, email_layout_xmlid='mail.mail_notification_light')
+                _logger.info("New device alert email sent for user <%s> to <%s>", user.login, user.email)
+
+    def _prepare_new_device_notice_values(self):
+        values = {
+            'date': format_datetime(self.env, datetime.now(), dt_format="long"),
```

Comment:
'date' is a bit vague (and impossible to grep if someday we want to cleanup context usage), so would have a more greppable-friendly variable like 'now_datetime' (or 'now_datetime_formatted' if you keep the format)